pkg/runtime/pubsub/bulkpub.go

```diff
@@ -0,0 +1,132 @@
+// It is used when the component does not implement BulkPublisher.
+type defaultBulkPublisher struct {
+	p contribPubsub.PubSub
+}
+
+// NewDefaultBulkPublisher returns a new defaultBulkPublisher from a PubSub.
+func NewDefaultBulkPublisher(p contribPubsub.PubSub) *defaultBulkPublisher {
+	return &defaultBulkPublisher{
+		p: p,
+	}
+}
+
+// BulkPublish publishes a list of messages to a topic as individual Publish requests.
+// If 'bulkPublishSerially' metadata is set to true, the messages are sent to the broker serially,
+// in the same order. Otherwise they are sent to the broker as parallel Publish requests.
+func (p *defaultBulkPublisher) BulkPublish(_ context.Context, req *contribPubsub.BulkPublishRequest) (contribPubsub.BulkPublishResponse, error) {
+	if req.Metadata[bulkPublishSeriallyKey] == "true" {
+		return p.bulkPublishSerial(req)
+	} else {
+		return p.bulkPublishParallel(req)
+	}
+}
+
+// bulkPublishSerial publishes messages in a serial order. This is slower, but ensures
+// that messages are sent to the broker in the same order as specified in the request.
+func (p *defaultBulkPublisher) bulkPublishSerial(req *contribPubsub.BulkPublishRequest) (contribPubsub.BulkPublishResponse, error) {
+	statuses := make([]contribPubsub.BulkPublishResponseEntry, 0, len(req.Entries))
+	hasError := false
+
+	for _, entry := range req.Entries {
+		status := p.bulkPublishSingleEntry(req.PubsubName, req.Topic, entry)
+		if status.Error != nil {
+			hasError = true
+		}
+
+		statuses = append(statuses, status)
+	}
+
+	var err error
+	if hasError {
+		err = ErrBulkPublishFailure
+	}
+
+	return contribPubsub.BulkPublishResponse{Statuses: statuses}, err
+}
+
+// bulkPublishParallel publishes messages in parallel. This is faster, but does not guarantee
+// that messages are sent to the broker in the same order as specified in the request.
+func (p *defaultBulkPublisher) bulkPublishParallel(req *contribPubsub.BulkPublishRequest) (contribPubsub.BulkPublishResponse, error) {
+	statuses := make([]contribPubsub.BulkPublishResponseEntry, 0, len(req.Entries))
+	var eg errgroup.Group
+	eg.SetLimit(utils.GetIntOrDefault(req.Metadata, bulkPublishMaxConcurrencyKey, defaultBulkPublishMaxConcurrency))
+
+	statusChan := make(chan contribPubsub.BulkPublishResponseEntry, len(req.Entries))
+
+	for _, entry := range req.Entries {
+		entry := entry
+		eg.Go(func() error {
+			status := p.bulkPublishSingleEntry(req.PubsubName, req.Topic, entry)
+			statusChan <- status
+			return status.Error
+		})
+	}
+
+	err := eg.Wait()
+	close(statusChan)
+
+	for status := range statusChan {
+		statuses = append(statuses, status)
+	}
+
+	return contribPubsub.BulkPublishResponse{Statuses: statuses}, err
+}
+
+// bulkPublishSingleEntry sends a single message to the broker as a Publish request.
+func (p *defaultBulkPublisher) bulkPublishSingleEntry(pubsubName, topic string, entry contribPubsub.BulkMessageEntry) contribPubsub.BulkPublishResponseEntry {
```

Comment:
Should this handle resiliency?

Reply:
Handling resiliency for bulk publish will be part of 1.10....

Reply:
Sounds good. I assume this is to give time to decide if we retry the individual events or the overall publish?